server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstPolicy.java

```diff
@@ -20,34 +20,78 @@
 package org.apache.druid.server.coordinator.duty;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Preconditions;
 import com.google.inject.Inject;
+import org.apache.druid.java.util.common.Pair;
 import org.apache.druid.server.coordinator.DataSourceCompactionConfig;
+import org.apache.druid.server.coordinator.DruidCoordinatorConfig;
 import org.apache.druid.timeline.SegmentTimeline;
 import org.joda.time.Interval;
 
+import java.time.Clock;
 import java.util.List;
 import java.util.Map;
 
 /**
  * This policy searches segments for compaction from the newest one to oldest one.
+ * The {@link #resetIfNeeded} functionality is inspired by {@link com.google.common.base.Suppliers.ExpiringMemoizingSupplier}.
  */
 public class NewestSegmentFirstPolicy implements CompactionSegmentSearchPolicy
 {
   private final ObjectMapper objectMapper;
+  private final long durationMillis;
+  private transient volatile NewestSegmentFirstIterator iterator;
```

Comment:
I think transient just means don't serialize this object because it is computed. This search policy shouldn't be serialized, but this pattern is used in Suppliers.memoizingSupplierWithExpiration, so I figured I'd stick with it since it is battle tested. I meant to add a comment providing credit for the inspiration of this pattern - I'll update the docs

Reply:
I guess I agree that `transient` is technically harmless since we don't participate in `Serializable`, I was more thinking that it being there makes the reader wonder if we do or why its there (like i did), and since we don't use `transient` anywhere else in the code.

Guava has it because
```
static class MemoizingSupplier<T> implements Supplier<T>, Serializable
```
so it needs these things because it implements `Serializable`, but we do not.

Reply:
sgtm I'll remove the transient to reduce confusion